tracing/src/main/java/com/palantir/tracing/Tracer.java

```diff
@@ -525,17 +572,31 @@ private static void notifyObservers(Span span) {
         return span;
     }
 
-    private static Span toSpan(OpenSpan openSpan, Map<String, String> metadata, String traceId) {
-        return Span.builder()
+    private static <T> Span toSpan(OpenSpan openSpan, TagRecorder<? super T> tag, T state, String traceId) {
+        Span.Builder builder = Span.builder()
                 .traceId(traceId)
                 .spanId(openSpan.getSpanId())
                 .type(openSpan.type())
                 .parentSpanId(openSpan.getParentSpanId())
                 .operation(openSpan.getOperation())
                 .startTimeMicroSeconds(openSpan.getStartTimeMicroSeconds())
-                .durationNanoSeconds(System.nanoTime() - openSpan.getStartClockNanoSeconds())
-                .putAllMetadata(metadata)
-                .build();
+                .durationNanoSeconds(System.nanoTime() - openSpan.getStartClockNanoSeconds());
+        tag.record(SpanBuilderTagAdapter.INSTANCE, builder, state);
```

Comment:
So this is quite hard to validate and probably maintain (if you make the callers do this), but *I feel* like the contract you have with TagRecorder is that you should call #isEmpty()?

```
if (!tag.isEmpty(state)) {
            tag.record(SpanBuilderTagAdapter.INSTANCE, builder, state);    
        }
```

Reply:
You're right, we might as well do the empty check here.
If it is empty, it shouldn't cost anything to call `tag.record`. The `isEmpty` method allows us to quickly validate whether or not we need to keep track of the instance in some cases.